src/main/java/seedu/taskell/logic/commands/UndoCommand.java

```diff
@@ -4,42 +4,66 @@
 
 /**
- * Undo most recent executed command (add or delete only for now)
+ * Undo previously executed command (add or delete only for now)
+ * Note: only for current session only (meaning after app is closed, history will be cleared)
  */
 public class UndoCommand extends Command {
     private static final Logger logger = LogsCenter.getLogger(UndoCommand.class.getName());
     
     public static final String COMMAND_WORD = "undo";
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Undo most recent command.\n"
-            + "Example: " + COMMAND_WORD;
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Undo a previously executed command.\n"
+            + "Parameters: INDEX (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 3";
     
-    public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
-    public static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
-    public static final String MESSAGE_DUPLICATE_TASK = "This task already exists in the task manager";
-    public static final String MESSAGE_NO_TASK_TO_UNDO = "No add or delete commands available to undo.";
-
-    private static ArrayList<String> commandHistory;
+    private static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
+    private static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
+    
+    private static final String MESSAGE_DUPLICATE_TASK = "This task already exists in the task manager";
+    private static final String MESSAGE_NO_TASK_TO_UNDO = "No add or delete commands available to undo.";
+    private static final String MESSAGE_COMMAND_HISTORY_EMPTY = "No command history available for undo.";
+    private static final String MESSAGE_INVALID_INDEX = "Index is invalid";
+    
+    private static ArrayList<CommandHistory> commandHistoryList;
     
-    private static String mostRecentCommand;
-    private static Task mostRecentAddedTask;
-    private static Task mostRecentDeletedTask;
+    private int index;
+    private CommandHistory commandHistory;
     
-    public UndoCommand() {
+    public UndoCommand(int index) {
+        logger.info("Creating UndoCommand with index: " + index);
+        this.index = index;
+    }
+    
+    public static ArrayList<String> getListOfCommandHistoryText() {
+        assert commandHistoryList != null;
+        assert !commandHistoryList.isEmpty();
+        
+        ArrayList<String> list = new ArrayList<>();
+        for (CommandHistory history: commandHistoryList) {
+            list.add(history.getCommandText());
+        }
         
+        return list;
     }
     
     @Override
     public CommandResult execute() {
-        if (mostRecentCommand==null) {
-            return new CommandResult(String.format(MESSAGE_NO_TASK_TO_UNDO));
+        
+        if (commandHistoryList.isEmpty()) {
+            return new CommandResult(String.format(MESSAGE_COMMAND_HISTORY_EMPTY));
         }
```

Comment:
else if should process after '{'

![image](https://cloud.githubusercontent.com/assets/13817632/19630805/793647c2-99c4-11e6-9f8e-1d8d8048af81.png)